doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,102 +1,126 @@
-      always performed using setoid rewriting, even for Leibniz’s equality, so one
-      has to ``Import Setoid`` to use this variant.
+   :n:`{? @natural } {? {| ? | ! } }`
+     :n:`natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
+     is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number
+     of rewrites to perform.
 
-   .. tacv:: rewrite @term by @tactic
+     '?' (without :n:`@natural`) performs the rewrite as many times as possible
+     (possibly zero times).
+     This form never fails.  '!' (without :n:`@natural`) performs the rewrite as many
+     times as possible
+     and at least once.  The tactic fails if the requested number of rewrites can't
+     be performed.
 
-      Use tactic to completely solve the side-conditions arising from the
-      :tacn:`rewrite`.
+   :n:`@occurrences`
+     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
+     any of them can be rewritten.  If not specified, only the first occurrence
+     in the conclusion is replaced.
 
-   .. tacv:: rewrite {+, @orientation @term} {? in @ident }
+     If :n:`at @occs_nums` is specified, rewriting is always done with setoid
+     rewriting, even for Leibniz’s equality.
 
-      Is equivalent to the `n` successive tactics :n:`{+; rewrite @term}`, each one
-      working on the first subgoal generated by the previous one. An :production:`orientation`
-      ``->`` or ``<-`` can be inserted before each :token:`term` to rewrite.  One
-      unique clause can be added at the end after the keyword in; it will then
-      affect all rewrite operations.
+     .. todo not finding occurrences within forall is not a general limitation of
+        occurrences.  Probably there are other tactics that have this limitation.
 
-   In all forms of rewrite described above, a :token:`term` to rewrite can be
-   immediately prefixed by one of the following modifiers:
+   :n:`by @ltac_expr3`
+     If specified, is used to resolve all side conditions generated by the tactic.
 
-   + `?` : the tactic :n:`rewrite ?@term` performs the rewrite of :token:`term` as many
-     times as possible (perhaps zero time). This form never fails.
-   + :n:`@natural?` : works similarly, except that it will do at most :token:`natural` rewrites.
-   + `!` : works as `?`, except that at least one rewrite should succeed, otherwise
-     the tactic fails.
-   + :n:`@natural!` (or simply :n:`@natural`) : precisely :token:`natural` rewrites of :token:`term` will be done,
-     leading to failure if these :token:`natural` rewrites are not possible.
+   .. exn:: Tactic failure: Setoid library not loaded.
+      :undocumented:
+
+      .. todo You can use Typeclasses Debug to tell whether rewrite used
+         setoid rewriting.  Example here: https://github.com/coq/coq/pull/13470#discussion_r539230973
+
+   .. exn:: The @term provided does not end with an equation.
+      :undocumented:
```

Comment:
After grepping "does not end" and "with an equation", I conclude that this error message doesn't exist anymore. The error I got while trying to apply something of the wrong type was "Cannot find a relation to rewrite."